### ISSUE TITLE
AWS auth configmap removes additional node group(s) after a second apply, causing additional node group(s) to fall into DEGRADED state

### ISSUE THREAD

**Issue by leptitchriss** [terraform-aws-modules/terraform-aws-eks#2289]
## Description

Basically the same thing as described here: https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1786

If the `manage_aws_auth_configmap` is set to `true` in the `eks` module (eg. to add authorized IAM users/roles):

```
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  (...)
  manage_aws_auth_configmap = true
```

... in addition to configuring an additional node group with the `eks_managed_node_group` module:

```
module "eks_managed_node_group" {
  source = "terraform-aws-modules/eks/aws//modules/eks-managed-node-group"
  (...)
```

On the first apply (bootstrap), everything works as expected. On a second apply, the additional node group config is removed from the configmap (example with placeholder values):

```
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.eks.kubernetes_config_map_v1_data.aws_auth[0] will be updated in-place
  ~ resource "kubernetes_config_map_v1_data" "aws_auth" {
      ~ data  = {
          ~ "mapRoles"    = <<-EOT
              - - groups:
              -   - system:bootstrappers
              -   - system:nodes
              -   rolearn: arn:aws:iam::<ACCOUNT>:role/<eks-module-node-group>
              -   username: system:node:{{EC2PrivateDNSName}}
              - - groups:
              -   - system:masters
              -   rolearn: arn:aws:iam::<ACCOUNT>:role/<some_role_for_users>
              -   username: <some_username>
              - - groups:
              -   - system:bootstrappers
              -   - system:nodes
              -   rolearn: arn:aws:iam::<ACCOUNT>:role/<additional-eks-node-group>
              -   username: system:node:{{EC2PrivateDNSName}}
              + - "groups":
              +   - "system:bootstrappers"
              +   - "system:nodes"
              +   "rolearn": "arn:aws:iam::<ACCOUNT>:role/<eks-module-node-group>"
              +   "username": "system:node:{{EC2PrivateDNSName}}"
              + - "groups":
              +   - "system:masters"
              +   "rolearn": "arn:aws:iam::<ACCOUNT>:role/<some_role_for_users>"
              +   "username": "<some_username>"
            EOT
            # (2 unchanged elements hidden)
        }
        id    = "kube-system/aws-auth"
        # (1 unchanged attribute hidden)

        # (1 unchanged block hidden)
    }
```

For clarity, in the above snippet, the following entry is being removed:

```
              - - groups:
              -   - system:bootstrappers
              -   - system:nodes
              -   rolearn: arn:aws:iam::<ACCOUNT>:role/<additional-eks-node-group>
              -   username: system:node:{{EC2PrivateDNSName}}
```

## Versions

- Module version [Required]: 18.29.0 +

- Terraform version: 1.1.5 +

## Additional context

Perhaps I am missing something, but is there a way to get this to work (eg. add additional node groups after bootstrapping the EKS cluster and not lose the auth configmap entries on subsequent `terraform apply`)?

**Comment by leptitchriss** [comment]
Looking at the code it doesn't look like this use-case was ever really considered: https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/main.tf#L386-L491

As a workaround I need to manually add the expected entry of the additional node group(s) to the `aws_auth_roles` section in the `eks` module:

```
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  (...)
  manage_aws_auth_configmap = true
  aws_auth_roles = [
    {
        rolearn = "arn:aws:iam::<ACCOUNT>:role/<additional-eks-node-group>"
        username = "system:node:{{EC2PrivateDNSName}}"
        groups = ["system:bootstrappers", "system:nodes"]
    },
    (...)
  ]
```

**Comment by leptitchriss** [comment]
I understand the complexity around this... but this is not ideal.

Although the initial entry into this configmap is done by the `eks_managed_node_group` (upon node group instantiation), I believe the `eks` module itself would need to be modified in some way to preserve these additional entries to keep `terraform apply` executions idempotent between runs (eg. do not constantly remove + add the entry back with each plan/apply).

**Comment by leptitchriss** [comment]
But as it currently stands, the entry is never re-added and causes the additional node groups to go into a DEGRADED state until the entry is added back with the above workaround.

**Comment by bryantbiggs** [comment]
the core module only knows about the resource it creates - such as IAM roles created for the EKS managed node groups, self-managed node groups, Fargate profiles. If you create node groups outside of the core module (standalone if you will), you need to add the roles for those node groups back to the aws-auth configmap of the core module. 

this isn't clear in our current examples so I will update the example to better shown this

**Comment by antonbabenko** [comment]
This issue has been resolved in version 18.31.0 :tada:
